### PR TITLE
Enable GeneratedRegex on partial properties

### LINKED ISSUE #77797: Update GeneratedRegex to support partial properties
Partial properties are destined for support in C# 13.  When that happens, we should utilize them in the regex generator (we wanted this experience from the get-go).

This will entail:
- Updating `GeneratedRegexAttribute` to target properties:
```diff
-[AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+[AttributeUsage(AttributeTargets.Method | AttributeTargets.Property, AllowMultiple = false, Inherited = false)]
 public sealed class GeneratedRegexAttribute : Attribute
```

- Updating the generator to support instance and static get-only properties, e.g. if a user writes:
```C#
[GeneratedRegex("...")]
private partial Regex MyProp1 { get; }

[GeneratedRegex("...")]
public static partial Regex MyProp2 { get; }
```
the generator emits:
```C#
private partial global::System.Text.RegularExpressions.Regex MyProp1 => global::System.Text.RegularExpressions.Generated.MyProp1_0.Instance;
public static partial global::System.Text.RegularExpressions.Regex MyProp2 => global::System.Text.RegularExpressions.Generated.MyProp2_0.Instance;
```

We will only support get-only auto props. Others will result in an error diagnostic.

### PR DESCRIPTION
**Pull request by stephentoub**
Extends the attribute to apply to properties, and updates the source generator to support:
```C#
[GeneratedRegex("...")]
private static partial Regex Prop { get; }
```
in addition to the already supported:
```C#
[GeneratedRegex("...")]
private static partial Regex Method();
```

Closes https://github.com/dotnet/runtime/issues/77797

### CODE REVIEW

**Review by RikkiGibson — APPROVED** [review]

File: src/libraries/System.Text.RegularExpressions/gen/Resources/Strings.resx [hunk 1]
```diff
@@ -131,8 +131,8 @@
   <data name="InvalidRegexArgumentsMessage" xml:space="preserve">
     <value>The specified regex is invalid. '{0}'</value>
   </data>
-  <data name="RegexMethodMustHaveValidSignatureMessage" xml:space="preserve">
-    <value>GeneratedRegexAttribute method must be partial, parameterless, non-generic, non-abstract, and return Regex.</value>
+  <data name="RegexMemberMustHaveValidSignatureMessage" xml:space="preserve">
+    <value>GeneratedRegexAttribute method or property must be partial, parameterless, non-generic, non-abstract, and return Regex. If a property, it must also be get-only.</value>
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/gen/Resources/Strings.resx:135]:
```suggestion
    <value>GeneratedRegexAttribute method or property must be partial, non-generic, non-abstract, and return Regex. If a method, it must also be parameterless. If a property, it must also be get-only.</value>
```
?

**Review by danmoseley — APPROVED** [review]

File: src/libraries/System.Text.RegularExpressions/gen/Resources/Strings.resx (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/gen/Resources/Strings.resx:135]: Indexers are properties that have parameters and are not supported.

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs [hunk 2]
```diff
@@ -217,18 +249,18 @@ SyntaxKind.RecordStructDeclaration or
...
             {
-                var location = method.GetLocation();
+                var location = syntax.GetLocation();
                 return Location.Create(location.SourceTree?.FilePath ?? string.Empty, location.SourceSpan, location.GetLineSpan().Span);
             }
         }
 
         /// <summary>Data about a regex directly from the GeneratedRegex attribute.</summary>
-        internal sealed record RegexPatternAndSyntax(RegexType DeclaringType, Location DiagnosticLocation, string MethodName, string Modifiers, string Pattern, RegexOptions Options, int? MatchTimeout, CultureInfo Culture, CompilationData CompilationData);
+        internal sealed record RegexPatternAndSyntax(RegexType DeclaringType, bool IsProperty, Location DiagnosticLocation, string MethodName, string Modifiers, bool NullableRegex, string Pattern, RegexOptions Options, int? MatchTimeout, CultureInfo Culture, CompilationData CompilationData);
```
Comment (campersau) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs:260]: Rename `MethodName` to `MemberName` here as well?

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs [hunk 3]
```diff
@@ -71,7 +71,12 @@ private static void EmitRegexPartialMethod(RegexMethod regexMethod, IndentedText
             writer.WriteLine($"/// </code>");
             writer.WriteLine($"/// </remarks>");
             writer.WriteLine($"[global::System.CodeDom.Compiler.{s_generatedCodeAttribute}]");
-            writer.WriteLine($"{regexMethod.Modifiers} global::System.Text.RegularExpressions.Regex {regexMethod.MethodName}() => global::{GeneratedNamespace}.{regexMethod.GeneratedName}.Instance;");
+            writer.Write($"{regexMethod.Modifiers} global::System.Text.RegularExpressions.Regex{(regexMethod.NullableRegex ? "?" : "")} {regexMethod.MethodName}");
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:74]: Should the model name be updated as well?
```suggestion
            writer.Write($"{regexMethod.Modifiers} global::System.Text.RegularExpressions.Regex{(regexMethod.NullableRegex ? "?" : "")} {regexMethod.MemberName}");
```

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs [hunk 4]
```diff
@@ -96,16 +106,36 @@ public partial class RegexGenerator
...
-                regexMethodSymbol.IsAbstract ||
-                regexMethodSymbol.Parameters.Length != 0 ||
-                regexMethodSymbol.Arity != 0 ||
-                !SymbolEqualityComparer.Default.Equals(regexMethodSymbol.ReturnType, regexSymbol))
+            bool nullableRegex;
+            if (regexMemberSymbol is IMethodSymbol regexMethodSymbol)
             {
-                return new DiagnosticData(DiagnosticDescriptors.RegexMethodMustHaveValidSignature, GetComparableLocation(methodSyntax));
+                if (!regexMethodSymbol.IsPartialDefinition ||
+                    regexMethodSymbol.IsAbstract ||
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs:116]: Does the source generator support both instance and static methods?

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs [hunk 5]
```diff
@@ -96,16 +106,36 @@ public partial class RegexGenerator
...
+                IPropertySymbol regexPropertySymbol = (IPropertySymbol)regexMemberSymbol;
+                if (!memberSyntax.Modifiers.Any(SyntaxKind.PartialKeyword) || // TODO: Switch to using regexPropertySymbol.IsPartialDefinition when available
+                    regexPropertySymbol.IsAbstract ||
+                    regexPropertySymbol.SetMethod is not null ||
+                    !SymbolEqualityComparer.Default.Equals(regexPropertySymbol.Type, regexSymbol))
+                {
+                    return new DiagnosticData(DiagnosticDescriptors.RegexMemberMustHaveValidSignature, GetComparableLocation(memberSyntax));
+                }
+
+                nullableRegex = regexPropertySymbol.NullableAnnotation == NullableAnnotation.Annotated;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs:138]: The value of `IPropertySymbol.NullableAnnotation` value does not take into account annotations deriving from attributes (`NotNullAttribute`, `MaybeNullAttribute`, etc) which need to be checked manually. We have helper code that does this in the STJ generator:

https://github.com/dotnet/runtime/blob/59e8bbcf83b664c3de6cfa553d9bbfad76578765/src/libraries/System.Text.Json/gen/Helpers/RoslynExtensions.cs#L301-L383

but adding a suppression at the generated expression is the quickest way to get guaranteed results.

**Review by eiriktsarpalis — APPROVED** [review]
Presumably this would require conceptual documentation updates?

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs (hunk 4 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs:116]: Yes

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs (hunk 5 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs:138]: That should not be necessary or desirable here. The only reason the nullable information is relevant is for the generated partial to match the manually-written partial. For annotations, that's relevant, or else the compiler may warn that they differ. But for attributes, we actually wouldn't want to duplicate them, as all attributes from across all partial declarations are accumulated, and if we duplicated them, we'd get compilation failures due to multiples not being allowed. I've augmented the tests to confirm that.

**Comment by stephentoub** [comment]
> Presumably this would require conceptual documentation updates?

Yes, we'll want to augment them to highlight the additional feature.

**Pull request merged**